File: Dynamic-Programming/best-time-to-buy-and-sell-stock.cpp
// Link: https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
class Solution {
public:
    // Time complexity: O(n)
    // Space complexity: O(1)
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        /*
        Key:  we need to find the minimum profit from day 0 --> day i
        - We know that maxProfit[i] = maxVal - minVal = prices[i] - minVal(0-->i)
        */
        int minimumProfit = INT_MAX;
        int maximumProfit = 0;
        for(int i = 0 ; i < n ; i++){
            minimumProfit = min(minimumProfit, prices[i]);
            maximumProfit = max(maximumProfit, prices[i] - minimumProfit);
        }
        return maximumProfit;
    }
};

File: Dynamic-Programming/best-time-to-buy-and-sell-stock.py
class Solution:
    # Time complexity: O(n)
    # Space complexity: O(1)
    def maxProfit(self, prices: List[int]) -> int:
        minProfit = float("inf")
        maxProfit = 0
        for p in prices:
            minProfit = min(minProfit, p)
            maxProfit = max(maxProfit, p - minProfit)
        return maxProfit